sync/sync_tail.go

```diff
@@ -0,0 +1,165 @@
+		if tail.IsZero() {
+			// no previously known Tail available - estimate solely on Head
+			estimatedHeight := estimateTail(head, s.Params.blockTime, s.Params.TrustingPeriod)
+			tail, err = s.getter.GetByHeight(ctx, estimatedHeight)
+			if err != nil {
+				return tail, fmt.Errorf("getting estimated tail(%d): %w", tailHeight, err)
+			}
+			fetched = true
+		} else {
+			// have a known Tail - estimate basing on it.
+			cutoffTime := head.Time().UTC().Add(-s.Params.PruningWindow)
+			diff := cutoffTime.Sub(tail.Time().UTC())
+			if diff <= 0 {
+				// current tail is relevant as is
+				return tail, nil
+			}
+			log.Debugw("current tail is beyond pruning window", "current_height", tail.Height(), "diff", diff.String())
+
+			toDeleteEstimate := uint64(diff / s.Params.blockTime) //nolint:gosec
+			estimatedNewTail := tail.Height() + toDeleteEstimate
+
+			for {
+				tail, err = s.store.GetByHeight(ctx, estimatedNewTail)
+				if err != nil {
+					log.Errorw("getting estimated tail from store", "height", estimatedNewTail, "error", err)
+					return tail, err
+				}
+				if tail.Time().UTC().Compare(cutoffTime) <= 0 {
+					// tail before or equal to cutoffTime
+					break
+				}
+
+				estimatedNewTail++
+			}
+
+			log.Debugw("estimated new tail", "new_height", tail.Height())
+		}
+	}
+
+	if fetched {
+		if err := s.store.Append(ctx, tail); err != nil {
+			return tail, fmt.Errorf("appending tail header: %w", err)
+		}
+	}
+
+	if err := s.moveTail(ctx, tail); err != nil {
+		return tail, fmt.Errorf("moving tail: %w", err)
+	}
+
+	return tail, nil
+}
+
+// moveTail moves the Tail to be the given header.
+// It will prune the store if the new Tail is higher than the old one or
+// sync up if the new Tail is lower than the old one.
+func (s *Syncer[H]) moveTail(ctx context.Context, to H) error {
+	from, err := s.store.Tail(ctx)
+	if errors.Is(err, header.ErrEmptyStore) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	switch {
+	case from.Height() < to.Height():
+		log.Infof("move tail up from %d to %d, pruning the diff...", from.Height(), to.Height())
+		err := s.store.DeleteTo(ctx, to.Height())
+		if err != nil {
+			return fmt.Errorf(
+				"deleting headers up to newly configured tail(%d): %w",
+				to.Height(),
+				err,
+			)
+		}
+	case from.Height() > to.Height():
+		log.Infof("move tail down from %d to %d, syncing the diff...", from.Height(), to.Height())
+
+		// TODO(@Wondertan): This works but it assumes this code is only run before syncing routine starts.
+		//  If run after, it may race with other in prog syncs.
+		//  To be reworked by bsync.
+		err := s.doSync(ctx, to, from)
+		if err != nil {
+			return fmt.Errorf("syncing the diff between from and new tail: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func estimateTail[H header.Header[H]](
+	head H,
+	blockTime, trustingPeriod time.Duration,
+) (height uint64) {
+	headersToRetain := uint64(trustingPeriod / blockTime) //nolint:gosec
```

Comment:
this can be extracted into var, no? `retentionWindow`

afaics can't change trustingPeriod or blockTime in runtime

Reply:
I don't follow what do you mean by extracting here

Reply:
doesn't matter just a nit